zebROS_ws/src/controllers_2023/src/four_bar_controller.cpp

```diff
@@ -0,0 +1,461 @@
+        {
+            position_command_.writeFromNonRT(FourBarCommand_2023 (four_bar_joint_.getPosition()));
+        }
+        const FourBarCommand_2023 setpoint = *(position_command_.readFromRT());
+        four_bar_joint_.setCommand(setpoint.GetAngle());
+
+        //if we're not climbing, add an arbitrary feed forward to hold the four_bar up
+       
+        four_bar_joint_.setMotionAcceleration(motion_magic_acceleration);
+        four_bar_joint_.setMotionCruiseVelocity(motion_magic_velocity);
+        four_bar_joint_.setPIDFSlot(0);
+ 
+        four_bar_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+        four_bar_joint_.setDemand1Value(arb_feed_forward);
+    }
+    else
+    {
+
+        four_bar_joint_.setMode(hardware_interface::TalonMode_PercentOutput);
+        if ((ros::Time::now() - last_time_down_).toSec() < four_bar_zeroing_timeout)
+        {
+            // Not yet zeroed. Run the four_bar down slowly until the limit switch is set.
+            ROS_INFO_STREAM_THROTTLE(0.25, "Zeroing four_bar with percent output: "
+                                     << four_bar_zeroing_percent_output);
+            four_bar_joint_.setCommand(four_bar_zeroing_percent_output);
+        }
+        //check stream
+        else
+        {
+            // Stop moving to prevent motor from burning out
+            ROS_INFO_STREAM_THROTTLE(0.25, "FourBar timed out");
+            four_bar_joint_.setCommand(0);
+        }
+
+        // If not zeroed but enabled, check if the arm is moving down
+        if ((four_bar_joint_.getMode() == hardware_interface::TalonMode_Disabled) ||
+                (four_bar_joint_.getSpeed() < 0)) // TODO : param
+        {
+            // If moving down, or disabled and thus not expected to move down, reset the timer
+            last_time_down_ = ros::Time::now();
+        }
+    }
+    last_angle_ = four_bar_joint_.getPosition();
+    last_mode_ = four_bar_joint_.getMode();
+}
+
+
+
+
+
+void FourBarController_2023::stopping(const ros::Time &/*time*/)
+{
+}
+
+//Command Service Function
+bool FourBarController_2023::cmdService(controllers_2023_msgs::FourBarSrv::Request  &req,
+        controllers_2023_msgs::FourBarSrv::Response &/*response*/)
+{   
+    if (req.position > max_extension_)
+    {
+        ROS_ERROR_STREAM("FourBar controller: req.position too forward : " << req.position << ". Stop violating physics! Setting to maximum position.");
+        req.position = max_extension_;
+    }
+    if (req.position < min_extension_)
+    {
+        ROS_ERROR_STREAM("FourBar controller: req.position too backward : " << req.position << ". Stop violating physics! Setting to minimum position.");
+        req.position = min_extension_;
+    }
+
+    if (isRunning())
+    {
+        //adjust talon mode, arb feed forward, and PID slot appropriately
+        double calcAngle = angleFromX(req.position, req.below);
+        if (isnan(calcAngle)) {
```

Comment:
Might be better to check !std::isfinite() instead

Reply:
oh cool, i didn't know about that function